// azure-kusto-data/dist-esm/src/client.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KustoClient = void 0;
const core_util_1 = require("@azure/core-util");
const axios_1 = __importDefault(require("axios"));
const http_1 = __importDefault(require("http"));
const https_1 = __importDefault(require("https"));
const uuid_1 = require("uuid");
const clientRequestProperties_1 = __importDefault(require("./clientRequestProperties"));
const cloudSettings_1 = __importDefault(require("./cloudSettings"));
const connectionBuilder_1 = __importDefault(require("./connectionBuilder"));
const errors_1 = require("./errors");
const kustoTrustedEndpoints_1 = require("./kustoTrustedEndpoints");
const response_1 = require("./response");
const security_1 = __importDefault(require("./security"));
const timeUtils_1 = require("./timeUtils");
const COMMAND_TIMEOUT_IN_MILLISECS = (0, timeUtils_1.toMilliseconds)(0, 10, 30);
const QUERY_TIMEOUT_IN_MILLISECS = (0, timeUtils_1.toMilliseconds)(0, 4, 30);
const CLIENT_SERVER_DELTA_IN_MILLISECS = (0, timeUtils_1.toMilliseconds)(0, 0, 30);
const MGMT_PREFIX = ".";
var ExecutionType;
(function (ExecutionType) {
    ExecutionType["Mgmt"] = "mgmt";
    ExecutionType["Query"] = "query";
    ExecutionType["Ingest"] = "ingest";
    ExecutionType["QueryV1"] = "queryv1";
})(ExecutionType || (ExecutionType = {}));
class KustoClient {
    constructor(kcsb) {
        this.cancelToken = axios_1.default.CancelToken.source();
        this._isClosed = false;
        this.connectionString = typeof kcsb === "string" ? new connectionBuilder_1.default(kcsb) : kcsb;
        if (!this.connectionString.dataSource) {
            throw new Error("Cluster url is required");
        }
        const url = new URL(this.connectionString.dataSource);
        this.cluster = url.toString();
        if (this.cluster.endsWith("/")) {
            this.cluster = this.cluster.slice(0, -1);
        }
        this.defaultDatabase = this.connectionString.initialCatalog;
        this.endpoints = {
            [ExecutionType.Mgmt]: `${this.cluster}/v1/rest/mgmt`,
            [ExecutionType.Query]: `${this.cluster}/v2/rest/query`,
            [ExecutionType.Ingest]: `${this.cluster}/v1/rest/ingest`,
            [ExecutionType.QueryV1]: `${this.cluster}/v1/rest/query`,
        };
        this.aadHelper = new security_1.default(this.connectionString);
        let headers = {
            Accept: "application/json",
        };
        if (core_util_1.isNode) {
            headers = Object.assign(Object.assign({}, headers), { "Accept-Encoding": "gzip,deflate", Connection: "Keep-Alive" });
        }
        const axiosProps = {
            headers,
            validateStatus: (status) => status === 200,
            maxBodyLength: Infinity,
            maxContentLength: Infinity,
            maxRedirects: 0,
        };
        // http and https are Node modules and are not found in browsers
        if (core_util_1.isNode) {
            // keepAlive pools and reuses TCP connections, so it's faster
            axiosProps.httpAgent = new http_1.default.Agent({ keepAlive: true });
            axiosProps.httpsAgent = new https_1.default.Agent({ keepAlive: true });
        }
        axiosProps.cancelToken = this.cancelToken.token;
        this.axiosInstance = axios_1.default.create(axiosProps);
    }
    async execute(db, query, properties) {
        query = query.trim();
        if (query.startsWith(MGMT_PREFIX)) {
            return this.executeMgmt(db, query, properties);
        }
        return this.executeQuery(db, query, properties);
    }
    async executeQuery(db, query, properties) {
        return this._execute(this.endpoints[ExecutionType.Query], ExecutionType.Query, db, { query }, properties);
    }
    async executeQueryV1(db, query, properties) {
        return this._execute(this.endpoints[ExecutionType.QueryV1], ExecutionType.QueryV1, db, { query }, properties);
    }
    async executeMgmt(db, query, properties) {
        return this._execute(this.endpoints[ExecutionType.Mgmt], ExecutionType.Mgmt, db, { query }, properties);
    }
    async executeStreamingIngest(db, table, stream, streamFormat, mappingName, blob, clientRequestId) {
        let endpoint = `${this.endpoints[ExecutionType.Ingest]}/${this.getDb(db)}/${table}?streamFormat=${streamFormat}`;
        if (mappingName != null) {
            endpoint += `&mappingName=${mappingName}`;
        }
        if (blob) {
            endpoint += `&sourceKind=uri`;
        }
        let properties = null;
        if (clientRequestId) {
            properties = new clientRequestProperties_1.default();
            properties.clientRequestId = clientRequestId;
        }
        return this._execute(endpoint, ExecutionType.Ingest, db, blob ? { blob } : { stream }, properties);
    }
    async _execute(endpoint, executionType, db, entity, properties) {
        this.ensureOpen();
        kustoTrustedEndpoints_1.kustoTrustedEndpoints.validateTrustedEndpoint(endpoint, (await cloudSettings_1.default.getCloudInfoForCluster(this.cluster)).LoginEndpoint);
        db = this.getDb(db);
        const headers = {};
        let payload;
        let clientRequestPrefix = "";
        const timeout = this._getClientTimeout(executionType, properties);
        let payloadContent = "";
        if ("query" in entity) {
            payload = {
                db,
                csl: entity.query,
            };
            if (properties != null) {
                payload.properties = properties.toJSON();
            }
            payloadContent = JSON.stringify(payload);
            headers["Content-Type"] = "application/json; charset=utf-8";
            clientRequestPrefix = "KNC.execute;";
        }
        else if ("stream" in entity) {
            payloadContent = entity.stream;
            clientRequestPrefix = "KNC.executeStreamingIngest;";
            if (core_util_1.isNode) {
                headers["Content-Encoding"] = "gzip";
                headers["Content-Type"] = "application/octet-stream";
            }
            else {
                headers["Content-Type"] = "application/json";
            }
        }
        else if ("blob" in entity) {
            payloadContent = {
                sourceUri: entity.blob,
            };
            clientRequestPrefix = "KNC.executeStreamingIngestFromBlob;";
            headers["Content-Type"] = "application/json";
        }
        else {
            throw new Error("Invalid parameters - expected query or streaming ingest");
        }
        let kustoHeaders = this.connectionString.clientDetails().getHeaders();
        kustoHeaders["x-ms-client-request-id"] = `${clientRequestPrefix}${(0, uuid_1.v4)()}`;
        if (properties != null) {
            kustoHeaders = Object.assign(Object.assign({}, kustoHeaders), properties.getHeaders());
        }
        for (const key of Object.keys(kustoHeaders)) {
            if (kustoHeaders[key]) {
                headers[key] = kustoHeaders[key];
            }
        }
        const authHeader = await this.aadHelper.getAuthHeader();
        if (authHeader != null) {
            headers.Authorization = authHeader;
        }
        return this._doRequest(endpoint, executionType, headers, payloadContent, timeout, properties);
    }
    getDb(db) {
        if (db == null) {
            if (this.defaultDatabase == null) {
                throw new Error("No database provided, and no default database specified in connection string");
            }
            db = this.defaultDatabase;
        }
        return db;
    }
    async _doRequest(endpoint, executionType, headers, payload, timeout, properties) {
        var _a, _b, _c;
        // replace non-ascii characters with ? in headers
        for (const key of Object.keys(headers)) {
            headers[key] = headers[key].replace(/[^\x00-\x7F]+/g, "?");
        }
        const axiosConfig = {
            headers,
            timeout,
        };
        let axiosResponse;
        try {
            axiosResponse = await this.axiosInstance.post(endpoint, payload, axiosConfig);
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                // Since it's impossible to modify the error request object, the only way to censor the Authorization header is to remove it.
                error.request = undefined;
                if ((_a = error === null || error === void 0 ? void 0 : error.config) === null || _a === void 0 ? void 0 : _a.headers) {
                    error.config.headers.Authorization = "<REDACTED>";
                }
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                if ((_c = (_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.request) === null || _c === void 0 ? void 0 : _c._header) {
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                    error.response.request._header = "<REDACTED>";
                }
                if (error.response && error.response.status === 429) {
                    throw new errors_1.ThrottlingError("POST request failed with status 429 (Too Many Requests)", error);
                }
            }
            throw error;
        }
        return this._parseResponse(axiosResponse.data, executionType, properties, axiosResponse.status);
    }
    _parseResponse(response, executionType, properties, status) {
        const { raw } = properties || {};
        if (raw === true || executionType === ExecutionType.Ingest) {
            return response;
        }
        let kustoResponse = null;
        try {
            if (executionType === ExecutionType.Query) {
                kustoResponse = new response_1.KustoResponseDataSetV2(response);
            }
            else {
                kustoResponse = new response_1.KustoResponseDataSetV1(response);
            }
        }
        catch (ex) {
            throw new Error(`Failed to parse response ({${status}}) with the following error [${ex}].`);
        }
        if (kustoResponse.getErrorsCount().errors > 0) {
            throw new Error(`Kusto request had errors. ${kustoResponse.getExceptions()}`);
        }
        return kustoResponse;
    }
    _getClientTimeout(executionType, properties) {
        if (properties != null) {
            const clientTimeout = properties.getClientTimeout();
            if (clientTimeout) {
                return clientTimeout;
            }
            const serverTimeout = properties.getTimeout();
            if (serverTimeout) {
                return serverTimeout + CLIENT_SERVER_DELTA_IN_MILLISECS;
            }
        }
        return executionType === ExecutionType.Query || executionType === ExecutionType.QueryV1 ? QUERY_TIMEOUT_IN_MILLISECS : COMMAND_TIMEOUT_IN_MILLISECS;
    }
    close() {
        if (!this._isClosed) {
            this.cancelToken.cancel("Client Closed");
        }
        this._isClosed = true;
    }
    ensureOpen() {
        if (this._isClosed) {
            throw new Error("Client is closed");
        }
    }
}
exports.KustoClient = KustoClient;
exports.default = KustoClient;
