// azure-kusto-data/dist-esm/src/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeUtils = exports.MatchRule = exports.kustoTrustedEndpoints = exports.KustoResultTable = exports.KustoResultRow = exports.KustoResultColumn = exports.KustoResponseDataSet = exports.KustoDataErrors = exports.KustoConnectionStringBuilder = exports.CloudSettings = exports.ClientRequestProperties = exports.Client = void 0;
const client_1 = __importDefault(require("./client"));
exports.Client = client_1.default;
const clientRequestProperties_1 = __importDefault(require("./clientRequestProperties"));
exports.ClientRequestProperties = clientRequestProperties_1.default;
const cloudSettings_1 = __importDefault(require("./cloudSettings"));
exports.CloudSettings = cloudSettings_1.default;
const connectionBuilder_1 = __importDefault(require("./connectionBuilder"));
exports.KustoConnectionStringBuilder = connectionBuilder_1.default;
const KustoDataErrors = __importStar(require("./errors"));
exports.KustoDataErrors = KustoDataErrors;
const kustoTrustedEndpoints_1 = require("./kustoTrustedEndpoints");
Object.defineProperty(exports, "kustoTrustedEndpoints", { enumerable: true, get: function () { return kustoTrustedEndpoints_1.kustoTrustedEndpoints; } });
Object.defineProperty(exports, "MatchRule", { enumerable: true, get: function () { return kustoTrustedEndpoints_1.MatchRule; } });
const models_1 = require("./models");
Object.defineProperty(exports, "KustoResultColumn", { enumerable: true, get: function () { return models_1.KustoResultColumn; } });
Object.defineProperty(exports, "KustoResultRow", { enumerable: true, get: function () { return models_1.KustoResultRow; } });
Object.defineProperty(exports, "KustoResultTable", { enumerable: true, get: function () { return models_1.KustoResultTable; } });
const response_1 = require("./response");
Object.defineProperty(exports, "KustoResponseDataSet", { enumerable: true, get: function () { return response_1.KustoResponseDataSet; } });
const timeUtils_1 = require("./timeUtils");
const TimeUtils = { toMilliseconds: timeUtils_1.toMilliseconds };
exports.TimeUtils = TimeUtils;
